### PR TITLE
add httpd

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

File: roles/httpd/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    path: /etc/httpd/conf.d/welcome.conf
+    state: absent
+
+- name: Create cert directories
+  ansible.builtin.file:
+    path: "{{ item }}"
+    state: directory
+  loop:
+    - /etc/pki/katello/certs
+    - /etc/pki/katello/private
```
Comment (ehelms) [inline roles/httpd/tasks/main.yml:26]: I'm OK with this for parity, but I would love to change this location.
I suppose if we were to run Apache as a container these would be secrets and it'd be a moot point.

File: roles/httpd/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+---
+- name: Install Apache httpd
```
Comment (ehelms) [inline roles/httpd/tasks/main.yml:2]: We could use `ubi9/httpd-24` but based on the list here we'd have to do some minor customization of it?

File: roles/httpd/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      dest: "private/katello-apache.key"
+
+- name: Configure foreman-ssl vhost
+  ansible.builtin.template:
+    src: foreman-ssl-vhost.conf.j2
+    dest: /etc/httpd/conf.d/foreman-ssl.conf
+  vars:
+    pulp_api_backend: http://localhost:8080/
+    pulp_content_backend: http://localhost:8080/
+    foreman_backend: http://localhost:3000/
```
Comment (ehelms) [inline roles/httpd/tasks/main.yml:50]: Do you think we need to worry about making these variables or stick with a "static / opinionated" implementation?

File: roles/httpd/templates/foreman-ssl-vhost.conf.j2 [hunk 4]
```diff
@@ -0,0 +1,104 @@
+<VirtualHost *:443>
+ServerName {{ ansible_fqdn }}
```
Comment (ehelms) [inline roles/httpd/templates/foreman-ssl-vhost.conf.j2:2]:
```suggestion
  ServerName {{ ansible_fqdn }}
```

File: roles/httpd/tasks/main.yml (hunk 1 above)
Comment (evgeni) [reply roles/httpd/tasks/main.yml:26]: Tell me a good location and I'll move :)

File: roles/httpd/templates/foreman-ssl-vhost.conf.j2 [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    RequestHeader unset REMOTE_USER
+    RequestHeader set REMOTE_USER "%{SSL_CLIENT_S_DN_CN}s" env=SSL_CLIENT_S_DN_CN
+    ProxyPass {{ pulp_api_backend }}/v2/
+    ProxyPassReverse {{ pulp_api_backend }}/v2/
+  </Location>
+
+  ProxyPass /pulp/container/ {{ pulp_content_backend }}/pulp/container/
+  ProxyPassReverse /pulp/container/ {{ pulp_content_backend }}/pulp/container/
+
+  <Location "/pulp/isos">
```
Comment (ehelms) [inline roles/httpd/templates/foreman-ssl-vhost.conf.j2:40]: Reminds me that this doesn't hurt anything, but we should track removing it.

File: roles/httpd/templates/foreman-ssl-vhost.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    RequestHeader unset X-CLIENT-CERT
+    RequestHeader set X-CLIENT-CERT "%{SSL_CLIENT_CERT}s" env=SSL_CLIENT_CERT
+    ProxyPass {{ pulp_content_backend }}/pulp/content disablereuse=on timeout=600
+    ProxyPassReverse {{ pulp_content_backend }}/pulp/content
+  </Location>
+
+  <Location "/pulp/repos">
```
Comment (ehelms) [inline roles/httpd/templates/foreman-ssl-vhost.conf.j2:47]: Reminds me that this doesn't hurt anything, but we should track removing it.

File: roles/httpd/tasks/main.yml (hunk 3 above)
Comment (evgeni) [reply roles/httpd/tasks/main.yml:50]: I think I'd make them variables, yes.

File: roles/httpd/templates/foreman-ssl-vhost.conf.j2 (hunk 5 above)
Comment (evgeni) [reply roles/httpd/templates/foreman-ssl-vhost.conf.j2:40]: yeah, I literally copied over the 05-ssl.conf from an existing machine as a template :see_no_evil:

File: roles/httpd/tasks/main.yml (hunk 2 above)
Comment (ehelms) [reply roles/httpd/tasks/main.yml:2]: Normalization of services, configurations mostly.

File: roles/httpd/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+httpd_ssl_dir: /etc/pki/katello
```
Comment (ehelms) [inline roles/httpd/defaults/main.yml:1]: Ideas:

- /etc/pki/httpd
- /etc/pki/tls

File: roles/httpd/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+httpd_ssl_dir: /etc/pki/httpd
+httpd_pulp_api_backend: http://localhost:8080
+httpd_pulp_content_backend: http://localhost:8080
+httpd_foreman_backend: http://localhost:3000
```
Comment (ehelms) [inline roles/httpd/defaults/main.yml:4]: I read a little bit, but seemed that Unix sockets were unfriendly in his container environment with podman and definitely do not extend to Openshift. Is that your read on it as well?

Comment (evgeni) [reply roles/httpd/defaults/main.yml:4]: That's also my thinking here, yes.

**Review by ehelms — APPROVED** [review]

File: roles/httpd/defaults/main.yml (hunk 8 above)
Comment (ekohl) [reply roles/httpd/defaults/main.yml:4]: We have proxy issues (HTTP 502 IIRC) when we used HTTP that went away when we switched to UNIX sockets. We also saw those issues with proxies from content proxies that went away with HTTP/2. Does Puma support HTTP/2 or how can we avoid introducing those issues again?

Also, you can actually use socket activation with containers where systemd opens the socket on the host and passes the file descriptor to a container. I use that with gunicorn (on something that isn't Pulp) and works well for me. Should we investigate that?

File: tests/httpd_test.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    assert httpd.port(HTTP_PORT).is_reachable
+
+
+def test_https_port(host):
+    httpd = host.addr(HTTP_HOST)
+    assert httpd.port(HTTPS_PORT).is_reachable
+
+
+def test_https_foreman_ping(host):
+    cmd = host.run('curl --cacert /root/certificates/certs/ca.crt --silent --output /dev/null --write-out \'%{http_code}\' https://quadlet.example.com/api/v2/ping')
```
Comment (ekohl) [inline tests/httpd_test.py:23]: I'd like to see more fixtures, like the CA file and hostname.

File: roles/httpd/defaults/main.yml (hunk 8 above)
Comment (ehelms) [reply roles/httpd/defaults/main.yml:4]: Moved to https://github.com/ehelms/foreman-quadlet/issues/31 to allow continued discussion

**Pull request merged**